File: app/src/main/java/it/amonshore/comikkua/data/web/CmkWebDao.java
package it.amonshore.comikkua.data.web;

import androidx.paging.PagingSource;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;

@Dao
public abstract class CmkWebDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract Long insert(AvailableComics comics);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(AvailableComics... comics);

    @Query("DELETE FROM tavailablecomics")
    public abstract int deleteAll();

    @Transaction
    public void refresh(AvailableComics... comics) {
        deleteAll();
        insert(comics);
    }

    @Query("SELECT * FROM tavailablecomics " +
            "ORDER BY name COLLATE NOCASE ASC, publisher COLLATE NOCASE ASC, version")
    @Transaction
    public abstract PagingSource<Integer, AvailableComics> getAvailableComicsPagingSource();

    @Query("SELECT * FROM tavailablecomics WHERE" +
            "(name LIKE :likeName OR publisher LIKE :likeName) " +
            "ORDER BY name COLLATE NOCASE ASC, publisher COLLATE NOCASE ASC, version")
    @Transaction
    public abstract PagingSource<Integer, AvailableComics> getAvailableComicsPagingSource(String likeName);
}

File: app/src/main/java/it/amonshore/comikkua/data/web/CmkWebDao.kt
package it.amonshore.comikkua.data.web

import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Dao
interface CmkWebDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(comics: List<AvailableComics>)

    @Query("DELETE FROM tAvailableComics")
    suspend fun deleteAll()

    @Query(
        """
        SELECT * FROM tAvailableComics
        ORDER BY name COLLATE NOCASE ASC, publisher COLLATE NOCASE ASC, version
    """
    )
    suspend fun getAvailableComicsList(): List<AvailableComics>

    @Query(
        """
        SELECT a.* 
        FROM tAvailableComics a LEFT OUTER JOIN tComics t ON a.sourceId = t.sourceId
        WHERE 
        t.sourceId IS NULL OR t.selected = 0
        ORDER BY a.name COLLATE NOCASE ASC, a.publisher COLLATE NOCASE ASC, a.version
    """
    )
    fun getNotFollowedComicsFLow(): Flow<List<AvailableComics>>
}